visvis2/_wrappers.py

```diff
@@ -0,0 +1,204 @@
+import ctypes
+
+import wgpu
+import numpy as np
+
+
+def array_from_shader_type(spirv_type):
+    """ Get a numpy array object from a SpirV type from python-shader.
+    """
+    return np.asarray(spirv_type())
+
+
+# todo: can this be generic enough, keeping the GPU bits out / optional?
+
+# todo: Support for updating unmapped data. Use something like updateRange to do subBufferUpdate
+
+
+class BaseBufferWrapper:
+    """ A base buffer wrapper that can be implemented for numpy, ctypes arrays,
+    or any other kind of array.
+    """
+
+    def __init__(self, data=None, nbytes=None, usage=None, mapped=False):
+        self._data = data
+        if nbytes is not None:
+            self._nbytes = int(nbytes)
+        elif data is not None:
+            self._nbytes = self._nbytes_from_data(data)
+        else:
+            raise ValueError("Buffer must be instantiated with either data or nbytes.")
+        if isinstance(usage, int):
+            self._usage = usage
+        elif isinstance(usage, str):
+            usages = usage.upper().replace(",", " ").replace("|", " ").split()
+            assert usages
+            self._usage = 0
+            for usage in usages:
+                self._usage |= getattr(wgpu.BufferUsage, usage)
+        else:
+            self._usage = wgpu.BufferUsage.STORAGE | wgpu.BufferUsage.VERTEX
+            # raise ValueError("BufferWrapper usage must be int or str.")
```

Comment:
I guess the usage attribute depends on the renderer being used. Can we think of a way to allow for that? Can the bufferwrapper class transparently pass on the usage flag that is being passed to it? In other words: if it doesn't process it at all and doesn't have a default, it doesn't need to know the type of the usage flag, I think?

Reply:
Interesting thought. Will have to look into this.